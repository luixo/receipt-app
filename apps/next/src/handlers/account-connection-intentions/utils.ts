import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
