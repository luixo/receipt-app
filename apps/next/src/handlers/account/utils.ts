import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
