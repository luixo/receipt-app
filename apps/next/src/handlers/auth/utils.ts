import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
