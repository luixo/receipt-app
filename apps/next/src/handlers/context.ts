import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
