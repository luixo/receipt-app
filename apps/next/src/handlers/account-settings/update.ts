import { z } from "zod";

import type { SimpleUpdateObject } from "next-app/db/types";
import { authProcedure } from "next-app/handlers/trpc";

import { DEFAULT_ACCOUNT_SETTINGS } from "./get";

type SettingsUpdateObject = SimpleUpdateObject<"accountSettings">;

export const procedure = authProcedure
	.input(
		z.discriminatedUnion("type", [
			z.strictObject({
				type: z.literal("autoAcceptDebts"),
				value: z.boolean(),
			}),
		]),
	)
	.mutation(async ({ ctx, input }) => {
		const { database } = ctx;
		const updateObject: SettingsUpdateObject = {};
		switch (input.type) {
			case "autoAcceptDebts":
				updateObject.autoAcceptDebts = input.value;
		}
		const existingSettings = await database
			.selectFrom("accountSettings")
			.where("accountSettings.accountId", "=", ctx.auth.accountId)
			.executeTakeFirst();
		if (!existingSettings) {
			await database
				.insertInto("accountSettings")
				.values({
					accountId: ctx.auth.accountId,
					...DEFAULT_ACCOUNT_SETTINGS,
					...updateObject,
				})
				.executeTakeFirst();
		} else {
			await database
				.updateTable("accountSettings")
				.set(updateObject)
				.where("accountSettings.accountId", "=", ctx.auth.accountId)
				.executeTakeFirst();
		}
	});
