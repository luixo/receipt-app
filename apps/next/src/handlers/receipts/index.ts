import * as trpc from "@trpc/server";
import { router as previewsRouter } from "./previews";
import { router as itemsRouter } from "./items";
import { AuthorizedContext } from "../context";

export const router = trpc
	.router<AuthorizedContext>()
	.merge(previewsRouter)
	.merge(itemsRouter);
