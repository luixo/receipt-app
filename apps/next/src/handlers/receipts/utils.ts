import { Selection } from "kysely";
import { z } from "zod";

import {
	ReceiptsSelectExpression,
	ReceiptsDatabase,
	Database,
} from "next-app/db";
import { AccountsId, ReceiptsId } from "next-app/db/models";
import { roleSchema } from "next-app/handlers/validation";

export type Role = z.infer<typeof roleSchema>;

export const getAccessRole = async (
	database: Database,
	receipt: { ownerAccountId: string; id: string },
	accountId: string
): Promise<Role | undefined> => {
	if (receipt.ownerAccountId === accountId) {
		return "owner";
	}
	const participant = await database
		.selectFrom("receiptParticipants")
		.innerJoin("users", (jb) =>
			jb.onRef("users.id", "=", "receiptParticipants.userId")
		)
		.innerJoin("accounts", (jb) =>
			jb.onRef("accounts.id", "=", "users.connectedAccountId")
		)
		.where("accounts.id", "=", accountId)
		.where("receiptParticipants.receiptId", "=", receipt.id)
		.select("role")
		.executeTakeFirst();
	if (!participant) {
		return;
	}
	const parsed = roleSchema.safeParse(participant.role);
	if (!parsed.success) {
		// TODO: add database-level validation
		return;
	}
	return parsed.data;
};

export const getReceiptById = <SE extends ReceiptsSelectExpression<"receipts">>(
	database: Database,
	id: ReceiptsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "receipts", SE> | undefined> =>
	database
		.selectFrom("receipts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();

export const getOwnReceipts = (
	database: Database,
	ownerAccountId: AccountsId
) =>
	database.selectFrom("receipts").where("ownerAccountId", "=", ownerAccountId);

export const getForeignReceipts = (
	database: Database,
	ownerAccountId: AccountsId
) =>
	database
		.selectFrom("users")
		.where("users.connectedAccountId", "=", ownerAccountId)
		.where("users.ownerAccountId", "<>", ownerAccountId)
		.innerJoin("receiptParticipants", (jb) =>
			jb.onRef("receiptParticipants.userId", "=", "users.id")
		)
		.innerJoin("receipts", (jb) =>
			jb.onRef("receipts.id", "=", "receiptParticipants.receiptId")
		);
