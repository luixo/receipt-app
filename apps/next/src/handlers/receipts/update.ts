import * as trpc from "@trpc/server";
import { MutationObject, sql } from "kysely";
import { z } from "zod";

import { receiptNameSchema } from "app/utils/validation";
import { getDatabase, Database } from "next-app/db";
import { ReceiptsDatabase } from "next-app/db/types";
import { getReceiptById } from "next-app/handlers/receipts/utils";
import { authProcedure } from "next-app/handlers/trpc";
import {
	currencyCodeSchema,
	receiptIdSchema,
} from "next-app/handlers/validation";

export const procedure = authProcedure
	.input(
		z.strictObject({
			id: receiptIdSchema,
			update: z.discriminatedUnion("type", [
				z.strictObject({
					type: z.literal("name"),
					name: receiptNameSchema,
				}),
				z.strictObject({ type: z.literal("issued"), issued: z.date() }),
				z.strictObject({ type: z.literal("locked"), value: z.boolean() }),
				z.strictObject({
					type: z.literal("currencyCode"),
					currencyCode: currencyCodeSchema,
				}),
			]),
		})
	)
	.mutation(async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const receipt = await getReceiptById(database, input.id, [
			"ownerAccountId",
			"lockedTimestamp",
		]);
		if (!receipt) {
			throw new trpc.TRPCError({
				code: "PRECONDITION_FAILED",
				message: `No receipt found by id ${input.id}`,
			});
		}
		if (receipt.ownerAccountId !== ctx.auth.accountId) {
			throw new trpc.TRPCError({
				code: "UNAUTHORIZED",
				message: `Receipt ${input.id} is not owned by ${ctx.auth.accountId}`,
			});
		}

		const updateTable = (
			localDatabase: Database,
			setObject: MutationObject<ReceiptsDatabase, "receipts", "receipts">
		) =>
			localDatabase
				.updateTable("receipts")
				.set(setObject)
				.where("id", "=", input.id)
				.executeTakeFirst();

		if (input.update.type === "locked") {
			if (receipt.ownerAccountId !== ctx.auth.accountId) {
				throw new trpc.TRPCError({
					code: "UNAUTHORIZED",
					message: `You don't have rights to change ${input.id} receipt locked state.`,
				});
			}
			const emptyItems = await database
				.selectFrom("receiptItems")
				.where("receiptItems.receiptId", "=", input.id)
				.leftJoin("itemParticipants", (qb) =>
					qb.onRef("itemParticipants.itemId", "=", "receiptItems.id")
				)
				.groupBy("receiptItems.id")
				.having(
					database.fn.sum<string>("itemParticipants.part"),
					"is",
					sql`null`
				)
				.select("receiptItems.id")
				.executeTakeFirst();
			if (emptyItems) {
				throw new trpc.TRPCError({
					code: "FORBIDDEN",
					message: `Receipt ${input.id} has items with no participants.`,
				});
			}
			if (input.update.value) {
				if (receipt.lockedTimestamp) {
					throw new trpc.TRPCError({
						code: "CONFLICT",
						message: `Receipt ${input.id} is already locked.`,
					});
				}
				await updateTable(database, { lockedTimestamp: new Date() });
				return;
			}
			if (!receipt.lockedTimestamp) {
				throw new trpc.TRPCError({
					code: "CONFLICT",
					message: `Receipt ${input.id} is not locked.`,
				});
			}
			await updateTable(database, { lockedTimestamp: null });
			return;
		}
		if (receipt.lockedTimestamp) {
			throw new trpc.TRPCError({
				code: "FORBIDDEN",
				message: `Receipt ${input.id} cannot be updated while locked.`,
			});
		}
		let setObject: MutationObject<ReceiptsDatabase, "receipts", "receipts"> =
			{};
		switch (input.update.type) {
			case "currencyCode":
				setObject = { currencyCode: input.update.currencyCode };
				break;
			case "issued":
				setObject = { issued: input.update.issued };
				break;
			case "name":
				setObject = { name: input.update.name };
				break;
		}
		await updateTable(database, setObject);
	});
