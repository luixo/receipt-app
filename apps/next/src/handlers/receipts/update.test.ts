import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { createAuthContext } from "@tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
	insertItemParticipant,
	insertReceipt,
	insertReceiptItem,
	insertReceiptParticipant,
	insertUser,
} from "@tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "@tests/backend/utils/expect";
import type { TestContext } from "@tests/backend/utils/test";
import { test } from "@tests/backend/utils/test";
import type { TRPCMutationInput } from "app/trpc";
import { t } from "next-app/handlers/trpc";

import { procedure } from "./update";
import {
	getRandomCurrencyCode,
	verifyCurrencyCode,
	verifyIssued,
	verifyName,
	verifyReceiptId,
} from "./utils.test";

const router = t.router({ procedure });

const updateDescribes = (
	getUpdate: () => TRPCMutationInput<"receipts.update">["update"],
	isLockedTimestampUpdated: (lockedBefore: boolean) => boolean,
	getNextTimestamp: () => Date | undefined = () => new Date(),
) => {
	const runTest = async ({
		ctx,
		lockedBefore,
	}: {
		ctx: TestContext;
		lockedBefore: boolean;
	}) => {
		const { sessionId, accountId } = await insertAccountWithSession(ctx);
		const { id: receiptId } = await insertReceipt(
			ctx,
			accountId,
			lockedBefore ? { lockedTimestamp: new Date("2020-06-01") } : undefined,
		);
		const { id: userId } = await insertUser(ctx, accountId);
		const { id: receiptItemId } = await insertReceiptItem(ctx, receiptId);
		await insertReceiptParticipant(ctx, receiptId, userId);
		await insertItemParticipant(ctx, receiptItemId, userId);

		// Verify unrelated data doesn't affect the result
		const { id: anotherUserId } = await insertUser(ctx, accountId);
		const { id: anotherReceiptId } = await insertReceipt(ctx, accountId);
		await insertReceiptParticipant(ctx, anotherReceiptId, anotherUserId);
		await insertReceiptItem(ctx, anotherReceiptId);

		const { id: foreignAccountId } = await insertAccount(ctx);
		const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);
		const { id: foreignReceiptId } = await insertReceipt(ctx, foreignAccountId);
		await insertReceiptParticipant(ctx, foreignReceiptId, foreignUserId);
		await insertReceiptItem(ctx, foreignReceiptId);

		const caller = router.createCaller(createAuthContext(ctx, sessionId));
		const result = await expectDatabaseDiffSnapshot(ctx, () =>
			caller.procedure({ id: receiptId, update: getUpdate() }),
		);
		expect(result).toStrictEqual<typeof result>(
			isLockedTimestampUpdated(lockedBefore)
				? { lockedTimestamp: getNextTimestamp() }
				: {},
		);
	};

	test("locked before update", async ({ ctx }) => {
		await runTest({ ctx, lockedBefore: true });
	});
	test("unlocked before update", async ({ ctx }) => {
		await runTest({ ctx, lockedBefore: false });
	});
};

describe("receipts.update", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure({
				id: faker.string.uuid(),
				update: { type: "name", name: faker.lorem.words() },
			}),
		);

		verifyReceiptId(
			(context, receiptId) =>
				router.createCaller(context).procedure({
					id: receiptId,
					update: { type: "name", name: faker.lorem.words() },
				}),
			"",
		);

		verifyName(
			(context, name) =>
				router.createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "name", name },
				}),
			"update.",
		);

		verifyIssued(
			(context, issued) =>
				router.createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "issued", issued },
				}),
			"update.",
		);

		verifyCurrencyCode(
			(context, currencyCode) =>
				router.createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "currencyCode", currencyCode },
				}),
			"update.",
		);

		test("receipt not found", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);

			// Verifying adding other receipts doesn't affect the error
			await insertReceipt(ctx, accountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const nonExistentReceiptId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						id: nonExistentReceiptId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"NOT_FOUND",
				`No receipt found by id "${nonExistentReceiptId}".`,
			);
		});

		test("receipt is not owned by the account", async ({ ctx }) => {
			// Self account
			const {
				sessionId,
				accountId,
				account: { email },
			} = await insertAccountWithSession(ctx);
			// Foreign account
			const { id: otherAccountId } = await insertAccount(ctx);
			const { id: foreignReceiptId } = await insertReceipt(ctx, otherAccountId);

			// Verifying adding other receipts doesn't affect the error
			await insertReceipt(ctx, accountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						id: foreignReceiptId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"FORBIDDEN",
				`Receipt "${foreignReceiptId}" is not owned by "${email}".`,
			);
		});

		test("receipt cannot be locked with items with no participants", async ({
			ctx,
		}) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: receiptId } = await insertReceipt(ctx, accountId);
			await insertReceiptItem(ctx, receiptId);

			// Verifying adding other receipts doesn't affect the error
			await insertReceipt(ctx, accountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						id: receiptId,
						update: { type: "locked", locked: true },
					}),
				"FORBIDDEN",
				`Receipt "${receiptId}" has items with no participants.`,
			);
		});
	});

	describe("functionality", () => {
		describe("update name", () => {
			updateDescribes(
				() => ({ type: "name", name: faker.lorem.words() }),
				() => false,
			);
		});

		describe("update issued", () => {
			updateDescribes(
				() => ({ type: "issued", issued: new Date("2020-06-01") }),
				() => false,
			);
		});

		describe("update currency code", () => {
			updateDescribes(
				() => ({ type: "currencyCode", currencyCode: getRandomCurrencyCode() }),
				(lockedBefore) => lockedBefore,
			);
		});

		describe("update locked - true", () => {
			updateDescribes(
				() => ({ type: "locked", locked: true }),
				() => true,
			);
		});

		describe("update locked - false", () => {
			updateDescribes(
				() => ({ type: "locked", locked: false }),
				() => true,
				() => undefined,
			);
		});
	});
});
