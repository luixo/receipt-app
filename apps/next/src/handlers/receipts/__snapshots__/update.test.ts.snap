// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`receipts.update > functionality > update currency code > locked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-06-01 00:00:00');
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'MAD', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-01-01 00:00:00');"
`;

exports[`receipts.update > functionality > update currency code > unlocked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'LTL', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);"
`;

exports[`receipts.update > functionality > update issued > locked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-06-01 00:00:00');
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-06-01 00:00:00', '2020-06-01 00:00:00');"
`;

exports[`receipts.update > functionality > update issued > unlocked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-06-01 00:00:00', NULL);"
`;

exports[`receipts.update > functionality > update locked - false > locked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-06-01 00:00:00');
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', NULL);"
`;

exports[`receipts.update > functionality > update locked - false > unlocked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);"
`;

exports[`receipts.update > functionality > update locked - true > locked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-06-01 00:00:00');
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-01-01 00:00:00');"
`;

exports[`receipts.update > functionality > update locked - true > unlocked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', '2020-01-01 00:00:00');"
`;

exports[`receipts.update > functionality > update name > locked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'aliquam corporis', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-06-01 00:00:00');
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('65b1d21a-4b9d-4019-a838-fbddad09c2b2', 'cupiditate asperiores aperiam', 'NOK', '2020-01-01 00:00:00', '16881856-073a-4f5f-84e4-969b374cba81', '2020-01-01 00:00:00', '2020-06-01 00:00:00');"
`;

exports[`receipts.update > functionality > update name > unlocked before update 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'fugit et', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);
@@ --- --- @@
+ INSERT INTO public.receipts VALUES ('ce2a0b7b-e615-42b9-be91-214ad04913ba', 'totam dolores amet', 'FKP', '2020-01-01 00:00:00', 'd8fbc0b6-dd9d-415a-ad00-7f2fbb3a992b', '2020-01-01 00:00:00', NULL);"
`;
